pyquil/analytics.py

```diff
@@ -0,0 +1,310 @@
+                gate_counts['JUMP-WHEN'] + \
+                gate_counts['JUMP-UNLESS']
+    results['Three+ Qubit Gates'] = \
+                gate_counts['CCNOT'] + \
+                gate_counts['CSWAP']
+    results['Elements in DEFGATES'] = \
+                gate_counts['DEFGATE']
+    return results
+
+def analyze(program_creator, input_list, should_range=False, to_include=None, remove_zeros=False):
+    '''
+    Analyzes a program by compiling properties of the pyQuil program for different inputs. Provides
+    a wide range of configurations to assist in analyzing specific programs
+    :param program_creator: A method to accept program parameters and create the Quil program (i.e
+                            a method such as qft from grove.fourier).
+    :param input_list: A list of inputs to be inputted into the program creator. These may be tuples
+                       of parameters, or something as simple as a list [0, 1, 2 ...] indicating the
+                       number of qubits to use in a program.
+    :param should_range: (Optional) If true, takes each value in n_list, assumes that it is an integer,
+                         and replaces it with range(that integer). This is useful in situations such as
+                         using 1, 2, 3, ... qubits without indicating the range of indices for these
+                         qubits. Default is False.
+    :param to_include: (Optional) A list of properties to analyze with your program creator. Default is
+                       to include every parameter.
+    :param remove_zeros: (Optional) If set to True, any properties where the value is 0 for all inputs
+                         of the program are left out of the final result. Default is False.
+    :return: A tuple containing multiple analysis results:
+                0) The input list used to generate each program
+                1) A structure of arrays, with indices corresponding to the index of the input in input_list,
+                and keys representing the properties measured for those program compilations
+                2) A dictionary from properties to possible complexities for that property (wrt input)
+    '''
+
+    results = []
+    for val in input_list:
+        # Space efficient option to pass in sizes of a list
+        # rather than a list of increasing lists. This is helpful
+        # for items such as qubit ranges
+        if should_range:
+            val = range(val)
+        
+        start_time = time.time() / 1000.0
+        program = program_creator(val)
+        properties = get_qubit_num_gates(get_gate_types(program))
+        properties["Total Instructions"] = get_total_instruction_count(program)
+        properties["Total Qubits"] = get_qubit_count(program)
+        properties["Compile Time"] = time.time() / 1000.0 - start_time
+        results.append(properties)
+        
+    results = aos_to_soa(results)
+    
+    # Filters results if specific ones are requested
```

Comment:
filtering can be a second step that is composed with this one rather than boolean flagged.  this is more modular